### PR TITLE
Get all string experiment

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
@rynowak here is my experiment in abstracting the management of resource strings properly.

I was wondering if maybe the interface should actually be for the strings instead of a stream since what we'd like to have as a final fix for this is a corefx API which would give us all the strings for a culture, rather than a stream.

### CODE REVIEW

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs [hunk 1]
```diff
@@ -50,7 +54,7 @@ public class ResourceManagerStringLocalizer : IStringLocalizer
         /// </summary>
         public ResourceManagerStringLocalizer(
             ResourceManager resourceManager,
-            AssemblyWrapper resourceAssemblyWrapper,
+            IResourceStreamManager resourceStreamManager,
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: Changing this would be breaking. But you should be able to convert `resourceAssemblyWrapper` to `AssemblyResourceStreamManager` and introduce a new ctor.

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStreamManager.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            if (!string.IsNullOrEmpty(culture.Name))
+            {
+                resourceStreamName += "." + culture.Name;
+            }
+            resourceStreamName += ".resources";
+
+            return resourceStreamName;
+        }
+
+        public string GetResourceStreamCacheKey(CultureInfo culture)
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStreamManager.cs]: We could return objects here (`CacheKey : IEquatable<CacheKey>`) and avoid creating a bunch of strings here.

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStreamManager.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            var cultureStartIndex = _assembly.FullName.IndexOf("Culture", StringComparison.OrdinalIgnoreCase);
+            if (cultureStartIndex < 0)
+            {
+                builder.Append(_assemblyElementDelimiter + cultureString);
+            }
+            else
+            {
+                var cultureEndIndex = _assembly.FullName.IndexOf(_assemblyElementDelimiter, cultureStartIndex);
+                var cultureLength = cultureEndIndex - cultureStartIndex;
+                builder = builder.Remove(cultureStartIndex, cultureLength);
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStreamManager.cs]: Don't need the assignments here.

**Comment by pranavkm** [comment]
Looks pretty nice

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs (hunk 1 above)
Comment (ryanbrandenburg) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: Unless I misunderstood him @rynowak was of the opinion that any ctor that had an internal type as a parameter (like `AssemblyWrapper`) was to be considered internal and we didn't have to worry too much about breaking it.

Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: Unfortunately the opposite could also be true: Since `ResourceManagerStringLocalizer` was public, we might need to consider every constructor parameter type public. Either way the `.Internal` stuff was a bit borked here.

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStreamManager.cs (hunk 2 above)
Comment (ryanbrandenburg) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStreamManager.cs]: But the public interface [here](https://github.com/aspnet/Localization/blob/dev/src/Microsoft.Extensions.Localization/IResourceNamesCache.cs#L20) defines it as a string so that would be a breaking API change. Unless I'm misunderstanding your ask?

File: src/Microsoft.Extensions.Localization/ResourceManagerWithCultureStringLocalizer.cs [hunk 4]
```diff
@@ -6,11 +6,12 @@
 using System.Globalization;
 using System.Reflection;
 using System.Resources;
+using Microsoft.Extensions.Localization.Internal;
 
 namespace Microsoft.Extensions.Localization
 {
     /// <summary>
-    /// An <see cref="IStringLocalizer"/> that uses the <see cref="System.Resources.ResourceManager"/> and
+    /// An <see cref="IStringLocalizer"/> that uses the <see cref="ResourceManager"/> and
```
Comment (dougbu) [inline src/Microsoft.Extensions.Localization/ResourceManagerWithCultureStringLocalizer.cs]: When non-design, do this fix for `ResourceReader` and in `ResourceManagerStringLocalizer`.

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs [hunk 5]
```diff
@@ -144,12 +168,29 @@ public class TestResourceManager : ResourceManager
             public TestResourceManager(string baseName, Assembly assembly)
                 : base(baseName, assembly)
             {
-
```
Comment (dougbu) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs]: :clap:

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStreamManager.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStreamManager.cs]: Nah, I thought it was using an `IMemoryCache` instance. Ignore this

**Comment by ryanbrandenburg** [comment]
:up: :date:

File: src/Microsoft.Extensions.Localization/Internal/IResourceStringManager.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
+﻿using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+
+namespace Microsoft.Extensions.Localization.Internal
+{
+    public interface IResourceStringManager
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/IResourceStringManager.cs]: `IResourceStringProvider`?

File: src/Microsoft.Extensions.Localization/Internal/IResourceStringManager.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    {
+        string GetResourceName(CultureInfo culture);
+
+        IList<string> GetAllResourceStrings(CultureInfo culture);
```
Comment (rynowak) [inline src/Microsoft.Extensions.Localization/Internal/IResourceStringManager.cs]: Your interface here exposes both the `resource name` concept and the `all strings` concept. Is there a way we could  relayer this code to get rid of `GetResourceName`? This isn't critical as a temporary/internal thing, this is already an improvement.

Comment (ryanbrandenburg) [inline src/Microsoft.Extensions.Localization/Internal/IResourceStringManager.cs]: I can accomplish this by making it the `IResourceStringManager`'s job to throw when it's missing the resources, it'll be in my next pass.

**Comment by ryanbrandenburg** [comment]
:up: :date:

**Comment by ryanbrandenburg** [comment]
Reminder for @pranavkm and @rynowak to give this a final look.

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+using System.IO;
+using System.Reflection;
+using System.Resources;
+using System.Text;
+
+namespace Microsoft.Extensions.Localization.Internal
+{
+    public class AssemblyResourceStringProvider : IResourceStringProvider
+    {
+        private const string _assemblyElementDelimiter = ", ";
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs]: `AssemblyElementDelimiter`

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        private static readonly string[] _assemblyElementDelimiterArray = new[] { _assemblyElementDelimiter };
+        private static readonly char[] _assemblyEqualDelimiter = new[] { '=' };
+
+        private AssemblyWrapper _assembly;
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs]: `readonly`?

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        }
+
+        private IList<string> ThrowOrNull(CultureInfo culture, bool throwOnMissing)
+        {
+            if (throwOnMissing)
+            {
+                throw new MissingManifestResourceException(
+                    Resources.FormatLocalization_MissingManifest(GetResourceName(culture)));
+            }
+            else
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs]: Super tiny nit, you could remove the `else` and just return here.

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+            return new AssemblyWrapper(assembly);
+        }
+
+        // This is all a workaround for https://github.com/dotnet/coreclr/issues/6123
+        private string ApplyCultureToAssembly(CultureInfo culture)
+        {
+            var builder = new StringBuilder(_assembly.FullName);
+
+            var cultureName = culture.Name == string.Empty ? "neutral" : culture.Name;
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs]: Is `culture.Name` ever null? Also `culture.Name.Length == 0`

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+            var cultureString = $"Culture={cultureName}";
+
+            var cultureStartIndex = _assembly.FullName.IndexOf("Culture", StringComparison.OrdinalIgnoreCase);
+            if (cultureStartIndex < 0)
+            {
+                builder.Append(_assemblyElementDelimiter + cultureString);
+            }
+            else
+            {
+                var cultureEndIndex = _assembly.FullName.IndexOf(_assemblyElementDelimiter, cultureStartIndex);
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs]: `StringComparison.Ordinal`

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+                var cultureLength = cultureEndIndex - cultureStartIndex;
+                builder.Remove(cultureStartIndex, cultureLength);
+                builder.Insert(cultureStartIndex, cultureString);
+            }
+
+            var firstSplit = _assembly.FullName.IndexOf(_assemblyElementDelimiter);
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs]: `assemblyNameEndIndex` or something like that?

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+            if (firstSplit < 0)
+            {
+                //Index of end of Assembly name
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs]: Space after `//`

File: src/Microsoft.Extensions.Localization/Internal/IResourceStringManager.cs [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace Microsoft.Extensions.Localization.Internal
+{
+    public interface IResourceStringProvider
+    {
+        IList<string> GetAllResourceStrings(CultureInfo culture, bool throwOnMissing);
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/IResourceStringManager.cs]: `IReadOnlyList<string>`?

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs [hunk 16]
```diff
@@ -50,7 +56,7 @@ public class ResourceManagerStringLocalizer : IStringLocalizer
         /// </summary>
         public ResourceManagerStringLocalizer(
             ResourceManager resourceManager,
-            AssemblyWrapper resourceAssemblyWrapper,
+            IResourceStringProvider resourceStringProvider,
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: Did we decide this wasn't breaking?

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs [hunk 17]
```diff
@@ -237,48 +236,9 @@ private IEnumerable<string> GetResourceNamesFromCultureHierarchy(CultureInfo sta
             return resourceNames;
         }
 
-        private string GetResourceStreamName(CultureInfo culture)
+        private IList<string> GetResourceNamesForCulture(CultureInfo culture, bool throwOnMissing)
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]:
```C#
private IList<string> GetResourceNamesForCulture(CultureInfo culture, bool throwOnMissing) => _resourceStringProvider.GetAllResourceStrings(culture, throwOnMissing);``` 

for :100: points

**Comment by pranavkm** [comment]
:shipit: from me. Have @rynowak sign off since he's assigned.

File: src/Microsoft.Extensions.Localization/Internal/IResourceStringManager.cs (hunk 15 above)
Comment (ryanbrandenburg) [inline src/Microsoft.Extensions.Localization/Internal/IResourceStringManager.cs]: That would require us to change `IResourceNamesCache`, which is a public interface.

File: src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+            return null;
+        }
+
+        public IList<string> GetAllResourceStrings(CultureInfo culture, bool throwOnMissing)
+        {
+            var assembly = GetAssembly(culture);
+            if (assembly == null)
+            {
+                return ThrowOrNull(culture, throwOnMissing);
+            }
```
Comment (rynowak) [inline src/Microsoft.Extensions.Localization/Internal/AssemblyResourceStringManager.cs]: Should this be inside the lambda?

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs (hunk 16 above)
Comment (rynowak) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: A constructor that uses `.Internal` types should be considered internal. @Eilon was happy with that principle when we discussed it.

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs [hunk 19]
```diff
@@ -237,48 +236,7 @@ private IEnumerable<string> GetResourceNamesFromCultureHierarchy(CultureInfo sta
...
-                        }
-                        return names;
-                    }
-                }
-
-            });
-
-            return cultureResourceNames;
-        }
+        private IList<string> GetResourceNamesForCulture(CultureInfo culture, bool throwOnMissing) =>
```
Comment (rynowak) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: Why do you even need this?

File: src/Microsoft.Extensions.Localization/ResourceManagerWithCultureStringLocalizer.cs [hunk 20]
```diff
@@ -20,9 +21,53 @@ public class ResourceManagerWithCultureStringLocalizer : ResourceManagerStringLo
         /// <summary>
         /// Creates a new <see cref="ResourceManagerWithCultureStringLocalizer"/>.
         /// </summary>
-        /// <param name="resourceManager">The <see cref="System.Resources.ResourceManager"/> to read strings from.</param>
+        /// <param name="resourceManager">The <see cref="ResourceManager"/> to read strings from.</param>
+        /// <param name="resourceStringProvider">The <see cref="IResourceStringProvider"/> that can find the resources.</param>
+        /// <param name="baseName">The base name of the embedded resource that contains the strings.</param>
+        /// <param name="resourceNamesCache">Cache of the list of strings for a given resource assembly name.</param>
+        /// <param name="culture">The specific <see cref="CultureInfo"/> to use.</param>
+        public ResourceManagerWithCultureStringLocalizer(
```
Comment (rynowak) [inline src/Microsoft.Extensions.Localization/ResourceManagerWithCultureStringLocalizer.cs]: Why is this here if it's going to expose an internal type

**Comment by rynowak** [comment]
:watch:

**Comment by ryanbrandenburg** [comment]
:up: :date:. Also noticed that I had renamed the class without renaming the file in 2 spots.

**Comment by rynowak** [comment]
:shipit:
